src/base64.cr

```diff
@@ -244,65 +244,89 @@ module Base64
+        bytes += 1
+      end
+      break if bytes > fin
 
-      yield (a << 2 | b >> 4).to_u8!
-      yield (b << 4 | c >> 2).to_u8!
-      yield (c << 6 | d).to_u8!
+      yield_decoded_chunk_bytes(
+        raise_error: true,
+        chunk_pos: bytes - bytes_begin,
+        bytes: {bytes[0], bytes[1], bytes[2], bytes[3]}
+      )
+      bytes += 4
     end
 
-    while (cstr < endcstr + 4) && (cstr.value == NL || cstr.value == NR)
-      cstr += 1
+    # Move the pointer by one byte until there is a valid base64 character or the end of `bytes` was reached
+    while (bytes < fin + 4) && (bytes.value == NL || bytes.value == NR)
+      bytes += 1
     end
 
-    mod = (endcstr - cstr) % 4
-    if mod == 2
-      a, b = next_decoded_value, next_decoded_value
-      yield (a << 2 | b >> 4).to_u8!
-    elsif mod == 3
-      a, b, c = next_decoded_value, next_decoded_value, next_decoded_value
-      yield (a << 2 | b >> 4).to_u8!
-      yield (b << 4 | c >> 2).to_u8!
-    elsif mod != 0
-      raise Error.new("Wrong size")
+    # If the amount of base64 characters is not divisable by 4, the remainder of the previous loop is handled here
+    unread_bytes = (fin - bytes) % 4
+    case unread_bytes
+    when 1
+      raise Base64::Error.new("Wrong size")
+    when 2
+      yield_decoded_chunk_bytes(
+        raise_error: true,
+        chunk_pos: bytes - bytes_begin,
+        bytes: {bytes[0], bytes[1]}
+      )
+    when 3
+      yield_decoded_chunk_bytes(
+        raise_error: true,
+        chunk_pos: bytes - bytes_begin,
+        bytes: {bytes[0], bytes[1], bytes[2]}
+      )
     end
   end
 
-  private macro next_decoded_value
-    sym = cstr.value
-    res = dt[sym]
-    cstr += 1
-    if res < 0
-      raise Error.new("Unexpected byte 0x#{sym.to_s(16)} at #{cstr - start_cstr - 1}")
-    end
-    res
+  # This macro decodes the given chunk of (2-4) base64 characters.
+  # The first argument (raise_error) determines if the macro raises an error when an invalid character is processed.
+  # The second argument (chunk_pos) is only used for the resulting error message.
+  # The resulting bytes are then each yielded to the given block, using the variable `byte`.
+  private macro yield_decoded_chunk_bytes(raise_error, chunk_pos, bytes, &block)
+    %buffer = 0_u32
+    {% for byte, i in bytes %}
+      %decoded = DECODE_TABLE.unsafe_fetch({{byte}})
+      %buffer = (%buffer << 6) + %decoded
+      {% if raise_error == true %}
+        raise Base64::Error.new("Unexpected byte 0x#{{{byte}}.to_s(16)} at #{{{chunk_pos}} + {{i}}}") if %decoded == 255_u8
+      {% end %}
+    {% end %}
+
+    # Each byte in the buffer is shifted to rightmost position of the buffer, then casted to a UInt8
+    {% for i in 2..(bytes.size) %}
+      yield (%buffer >> {{ (4 - bytes.size) * 2 + (8 * (bytes.size - i)) }}).to_u8!
+    {% end %}
   end
 
-  private DECODE_TABLE = Array(Int8).new(256) do |i|
+  private DECODE_TABLE = StaticArray(UInt8, 256).new do |i|
```

Comment:
What's the reason for this change from `Array` to `StaticArray`?

Reply:
Performance.
I need to benchmark this later, but it definitely gave a performance boost.

Reply:
Results:
```
staticarray 363.28k (  2.75µs) (± 4.41%)  2.42kB/op   1.06× slower
      array 384.81k (  2.60µs) (± 4.90%)  2.42kB/op        fastest
```

Reply:
Is the 6% improvement worth the 256 byte stack space or should it be changed back?

Reply:
I'm wondering where this difference is coming from... @asterite any thoughts what we should do here?

Reply:
Ideally, have a syntax for const arrays and declare this as such, so that it goes in the ROM section. Given that we don't have that, it doesn't matter.

Reply:
I actually didn't notice that when writing the comment.
I'll turn it back to an array later.